### PR TITLE
KAFKA-5515: Remove date formatting from Segments

### PR DESCRIPTION
**Pull request by dguy**
Remove date formatting from `Segments` and use the `segementId` instead.
Add tests to make sure can load old segments.
Rename old segment dirs to new formatting at load time.

### CODE REVIEW

**Comment by dguy** [comment]
@guozhangwang, @mjsax, @bbejeck

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java [hunk 1]
```diff
@@ -185,12 +186,31 @@ private void cleanup(final long segmentId) {
...
+            } catch (ParseException e) {
+                throw new ProcessorStateException("Unable to parse segmentName:"
+                                                          + segmentName
+                                                          + " to a date");
+            }
+            final long segmentId = date.getTime() / segmentInterval;
+            final File newName = new File(parent, segmentName(segmentId));
+            final File oldName = new File(parent, segmentName);
+            if (!oldName.renameTo(newName)) {
+                throw new ProcessorStateException("Unable to rename old style segment from: "
```
Comment (tedyu) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: If we don't throw exception here, future calls to segmentIdFromSegmentName would try to rename (and may succeed).
Is that better than throwing exception immediately ?

Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: There is no future call. This happens only when the database is being opened.

Comment (tedyu) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: Segments#getOrCreateSegment() would be called after the open.
If Segments class keeps a Set of segmentId's for which renameTo() was not successful, the old directory can still be used.
IMHO not being able to rename is not critical error. This should be rare case where the use of formatter has negligible impact on performance.

Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: It is not called from `Segments#getOrCreateSegment`

Comment (tedyu) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: segmentName() (where the format change happens) is called from Segments#getOrCreateSegment().

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java [hunk 2]
```diff
@@ -185,12 +186,31 @@ private void cleanup(final long segmentId) {
         }
     }
 
-    private long segmentIdFromSegmentName(String segmentName) {
-        try {
-            Date date = formatter.parse(segmentName.substring(name.length() + 1));
-            return date.getTime() / segmentInterval;
-        } catch (Exception ex) {
-            return -1L;
+    private long segmentIdFromSegmentName(String segmentName, final File parent) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: - add `final`
- single parameter per line formatting

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java [hunk 3]
```diff
@@ -185,12 +186,31 @@ private void cleanup(final long segmentId) {
...
+            final File oldName = new File(parent, segmentName);
+            if (!oldName.renameTo(newName)) {
+                throw new ProcessorStateException("Unable to rename old style segment from: "
+                                                          + oldName
+                                                          + " to new name: "
+                                                          + newName);
+            }
+            return segmentId;
+        } else {
+            return Long.parseLong(segmentName.substring(name.length() + 1));
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: Should we catch a parsing exceptions and throw `ProcessorStateException` instead?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java [hunk 4]
```diff
@@ -66,7 +67,7 @@ long segmentId(long timestamp) {
     }
 
     String segmentName(long segmentId) {
-        return name + "-" + formatter.format(new Date(segmentId * segmentInterval));
+        return name + ":" + segmentId;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: Should we add a comment about the old format to avoid the usage of `-` for another potential format change in the future?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java [hunk 5]
```diff
@@ -66,7 +67,7 @@ long segmentId(long timestamp) {
     }
 
     String segmentName(long segmentId) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: Not part of this PR, but add `final`

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java [hunk 6]
```diff
@@ -185,12 +186,31 @@ private void cleanup(final long segmentId) {
...
-            return -1L;
+    private long segmentIdFromSegmentName(String segmentName, final File parent) {
+        // old style segment name with date
+        if (segmentName.charAt(name.length()) == '-') {
+            final String datePart = segmentName.substring(name.length() + 1);
+            final Date date;
+            try {
+                date = formatter.parse(datePart);
+            } catch (ParseException e) {
+                throw new ProcessorStateException("Unable to parse segmentName:"
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: Before the change, we did return `-1L` but did not throw -- How does throwing an exception influence the exception path ?

**Review by mjsax — COMMENTED** [review]
Thanks for the PR. Couple of comments.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java (hunk 6 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: I was wondering why we returned `-1L`. I don't think it is valid, i.e, what are you actually opening? It could be an invalid directory in which case there would be an exception when the database opened. If it isn't invalid and there are multiple directories that fail with this then all the segments are `-1L` which isn't correct either. So... I thought it would be better to bail here as the state directory must have some invalid or tinkered with dirs in it

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: I agree that returning `-1L` does not seems to be correct. I am just wondering, if we handle this new exception properly as we change error handling with this change. Just try to understand the impact and if we need to do some other changes somewhere else or not.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java [hunk 7]
```diff
@@ -61,12 +62,14 @@ static long segmentInterval(long retentionPeriod, int numSegments) {
...
+    long segmentId(final long timestamp) {
         return timestamp / segmentInterval;
     }
 
-    String segmentName(long segmentId) {
-        return name + "-" + formatter.format(new Date(segmentId * segmentInterval));
+    String segmentName(final long segmentId) {
+        // previous format used - as a separator so if this changes in the future
+        // then we should use something different.
+        return name + ":" + segmentId;
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: Is this a non compatible change, since the segment name is used as the underlying store name, which is used for 1) changelog topic name, and 2) folder / file names under state dir.

Also we are losing a bit on the debugging convenience as the current store name contains the interval boundary itself.

**Comment by guozhangwang** [comment]
@bluemonk3y could you also take a look as you have done investigations on this issue also (https://github.com/bluemonk3y/kafka/commit/60176f037a8c1209f85597a69f1be9d833c6c637)?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java (hunk 7 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: The segment name isn't used for the changelog topic name. The changelog topic name derives purely from the `name` which hasn't changed.
The folder names will be different, but it can load the older data, so it is backward compatible in that respect.
For the debugability bit, if that is really a concern, we could change the name to be:
`name + ":" + (segmentId * segmentInterval)` ?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java (hunk 1 above)
Comment (bluemonk3y) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: File.renameTo can be unreliable on different platforms; dependent upon OS and FS. I've seen it fail erratically with no explanation and sometimes load related. The ugly fallback is to copy the file.

**Comment by bluemonk3y** [comment]
@guozhangwang dropping SDF is the right move and looks like @dguy addressed backward compatibility.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java (hunk 7 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: > For the debugability bit, if that is really a concern, we could change the name to be:
> name + ":" + (segmentId * segmentInterval) ?

Sounds good to me.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java (hunk 6 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: Originally if parse exception happens we simply do not load this segment at all, which is simply a bad design as a hindsight. This change is much better and we just need to remember handle the thrown exception in some higher call level.

However, I'm wondering if the folder would ever contain any other files besides the segment database folders like `.LOCK` files, etc? If that could be true, maybe we should log-and-skip than throwing or silently-skip.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java (hunk 7 above)
Comment (tedyu) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: The interval itself is computed from:
```
    static long segmentInterval(long retentionPeriod, int numSegments) {
        return Math.max(retentionPeriod / (numSegments - 1), MIN_SEGMENT_INTERVAL);
```
How about this formation:
```
name + ":" + segmentId + ":" + segmentInterval
```

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java (hunk 1 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: @tedyu I think @dguy 's point is that `segmentIdFromSegmentName()` is only called when we are opening the database and trying to load the existing segments, where we would like to try rename the segment names if they are in the old format (as you mentioned earlier that `future calls to segmentIdFromSegmentName` actually would not happen).

I'm wondering, if it is really worthy renaming the files? Since the older segments would be removed eventually as time advances, and after we have opened the segment database we would only reference them by ids and not by segment names any more, maybe we do not need to rename them at all?

**Review by guozhangwang — COMMENTED** [review]
Left two meta questions, otherwise looks good.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java (hunk 1 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: That is correct @guozhangwang. As far not renaming the segment: there was a reason i did it, but i'll have to take a look at why that was the case.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java (hunk 6 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: I'm not sure it would contain other files, but i guess it could. Maybe logging a warning and moving on is ok

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java (hunk 7 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: In my testing doing  `name + ":" + segmentId * segmentInterval` is faster than doing `name + ":" + segmentId + ":" + segmentInterval`

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java (hunk 1 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: @guozhangwang, the segment needs to be renamed as the segment name is derived in `getOrCreateSegment` Line 86, i.e., it does `Segment newSegment = new Segment(segmentName(segmentId), name, segmentId);`
at this point there is no way to tell if it was an old style name or a new style name.
Also, this renaming will only be a 1 off event, i.e., the first time it runs after an upgrade, so i feel it is fine to leave it as is.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java (hunk 7 above)
Comment (tedyu) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: Mind sharing how you benchmarked between the two alternatives ?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java (hunk 6 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: > I'm not sure it would contain other files, but i guess it could. Maybe logging a warning and moving on is ok

Cool

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java (hunk 1 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java]: Sounds Good

**Comment by guozhangwang** [comment]
retest this please

**Comment by guozhangwang** [comment]
LGTM. Please feel free to merge after it jenkins passed.

**Comment by dguy** [comment]
retest this please

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java [hunk 8]
```diff
@@ -148,6 +155,36 @@ public void shouldRollSegments() {
...
+
+        bytesStore = new RocksDBSegmentedBytesStore(storeName,
+                                                    retention,
+                                                    numSegments,
+                                                    schema);
+
+        bytesStore.init(context, bytesStore);
+        final List<KeyValue<Windowed<String>, Long>> results = toList(bytesStore.fetch(Bytes.wrap(key.getBytes()), 0L, 60000L));
+        assertThat(results, equalTo(Arrays.asList(KeyValue.pair(new Windowed<>(key, new SessionWindow(0L, 0L)), 50L),
+                                                  KeyValue.pair(new Windowed<>(key, new SessionWindow(30000L, 60000L)), 100L))));
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java:184]: nit: Could we also assert that the file names have been renamed to the new style as well? @dguy

**Comment by dguy** [comment]
merged to trunk
